### PR TITLE
Consume ACL for APIPA Endpoint from CreateNC Req

### PR DESCRIPTION
**Pull request by aegal**
**What this PR does / why we need it**:

This PR consumes endpoint policies targeting APIPA endpoints. In the future, we are extending policies to requested by the orchestrator in a extensible way from the Create Network Container Request. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #533 

**Special notes for your reviewer**:

Tested the successful creation of the HostNCApipaEndpoint with the additional policies requested. 
 

**Release note**:

```release-note
```

### CODE REVIEW

File: cns/NetworkContainerContract.go [hunk 1]
```diff
@@ -63,12 +68,14 @@ type CreateNetworkContainerRequest struct {
 	Routes                     []Route
 	AllowHostToNCCommunication bool
 	AllowNCToHostCommunication bool
+	EndpointPolicies           []NetworkContainerRequestPolicies
```
Comment (ashvindeodhar) [inline cns/NetworkContainerContract.go:71]: Should this be named as NetworkContainerPolicies?

File: cns/NetworkContainerContract.go [hunk 2]
```diff
@@ -63,12 +68,14 @@ type CreateNetworkContainerRequest struct {
...
 }
 
-// ConfigureContainerNetworkingRequest - specifies request to attach/detach container to network.
```
Comment (ashvindeodhar) [inline cns/NetworkContainerContract.go:68]: why did you remove ConfigureContainerNetworkingRequest struct?

File: cns/NetworkContainerContract.go [hunk 3]
```diff
@@ -220,3 +227,29 @@ type UnpublishNetworkContainerResponse struct {
...
+// Validate - Validates network container request policies
+func (networkContainerRequestPolicy *NetworkContainerRequestPolicies) Validate() error {
+
+	// validate ACL policy
+	if strings.EqualFold(networkContainerRequestPolicy.Type, "ACLPolicy") {
+
+		var requestedAclPolicy hcn.AclPolicySetting
+
+		if err := json.Unmarshal(networkContainerRequestPolicy.Settings, &requestedAclPolicy); err != nil {
+
```
Comment (ashvindeodhar) [inline cns/NetworkContainerContract.go:240]: remove blank lines

File: cns/NetworkContainerContract.go [hunk 4]
```diff
@@ -220,3 +227,29 @@ type UnpublishNetworkContainerResponse struct {
...
+			return fmt.Errorf("ACL policy failed to pass validation with error: %+v ", err)
+		}
+
+		if requestedAclPolicy.Action == "" {
```
Comment (ashvindeodhar) [inline cns/NetworkContainerContract.go:244]: Can you strip the spaces in the string and then check if the len is 0 ?

File: cns/NetworkContainerContract.go [hunk 5]
```diff
@@ -220,3 +227,29 @@ type UnpublishNetworkContainerResponse struct {
...
+			return errors.New("Action field cannot be empty in ACL Policy")
+		}
+
+		if requestedAclPolicy.Priority == 0 {
+			return errors.New("Priority field cannot be empty in ACL Policy")
```
Comment (ashvindeodhar) [inline cns/NetworkContainerContract.go:249]: I see you are creating error using fmt.Errorf on line 241 and errors.New in other places. Can you use either one of those in all the places? fmt.Errorf is what we have used (in most places) in the code

File: cns/hnsclient/hnsclient_windows.go [hunk 6]
```diff
@@ -426,6 +433,31 @@ func configureAclSettingHostNCApipaEndpoint(
...
 			}
 		}
+
+	}
+
+	if ncRequestedPolicies != nil {
+		// Iterate thru the requested endpoint policies where policy type is ACL, endpoint type is APIPA
+		// include the raw json message in the endpoint policies
+		for _, requestedPolicy := range ncRequestedPolicies {
+
```
Comment (ashvindeodhar) [inline cns/hnsclient/hnsclient_windows.go:443]: Remove blank lines in general here.

File: cns/hnsclient/hnsclient_windows.go [hunk 7]
```diff
@@ -426,6 +433,31 @@ func configureAclSettingHostNCApipaEndpoint(
...
+			if strings.EqualFold(requestedPolicy.Type, aclPolicyType) && strings.EqualFold(requestedPolicy.EndpointType, apipaEndpointType) {
```
Comment (ashvindeodhar) [inline cns/hnsclient/hnsclient_windows.go:444]: I believe you need to check if requestedPolicy is nil before accessing the members.

File: cns/hnsclient/hnsclient_windows.go [hunk 8]
```diff
@@ -426,6 +433,31 @@ func configureAclSettingHostNCApipaEndpoint(
...
+
+				var requestedAclPolicy hcn.AclPolicySetting
+
+				if err := json.Unmarshal(requestedPolicy.Settings, &requestedAclPolicy); err != nil {
```
Comment (ashvindeodhar) [inline cns/hnsclient/hnsclient_windows.go:448]: I think you don't need to create new err here. ( err := )

File: cns/hnsclient/hnsclient_windows.go [hunk 9]
```diff
@@ -490,7 +524,8 @@ func CreateHostNCApipaEndpoint(
 	networkContainerID string,
```
Comment (ashvindeodhar) [inline cns/hnsclient/hnsclient_windows.go:524]: you need to update the func definition signature in hnsclient_linux.go too.

File: cns/hnsclient/hnsclient_windows.go [hunk 10]
```diff
@@ -426,6 +433,31 @@ func configureAclSettingHostNCApipaEndpoint(
...
+			if strings.EqualFold(requestedPolicy.Type, aclPolicyType) && strings.EqualFold(requestedPolicy.EndpointType, apipaEndpointType) {
+
+				var requestedAclPolicy hcn.AclPolicySetting
+
+				if err := json.Unmarshal(requestedPolicy.Settings, &requestedAclPolicy); err != nil {
+
+					return nil, fmt.Errorf("Failed to Unmarshal requested ACL policy: %+v with error: %S", requestedPolicy.Settings, err)
+				}
+
+				logger.Printf("ACL Policy requested in NcGoalState %+v ", requestedAclPolicy)
```
Comment (ashvindeodhar) [inline cns/hnsclient/hnsclient_windows.go:453]: remove space after %+v

File: cns/NetworkContainerContract.go [hunk 11]
```diff
@@ -220,3 +227,29 @@ type UnpublishNetworkContainerResponse struct {
...
 	UnpublishResponseBody []byte
 }
+
+// Validate - Validates network container request policies
+func (networkContainerRequestPolicy *NetworkContainerRequestPolicies) Validate() error {
+
+	// validate ACL policy
+	if strings.EqualFold(networkContainerRequestPolicy.Type, "ACLPolicy") {
+
+		var requestedAclPolicy hcn.AclPolicySetting
```
Comment (ashvindeodhar) [inline cns/NetworkContainerContract.go:237]: This is specific to hnsv2 ( hcn.AclPolicySetting ). This may not work if the client is using docker and is specifying acl policy which is hnsv1 specific. This is not an issue at the moment but I think we should restrict the validation for EndpointType with APIPA for now. i.e. currently we only support this for APIPA endpoint type.
We can discuss more on this when we want to extend this to other endpoint too.

File: cns/NetworkContainerContract.go (hunk 2 above)
Comment (aegal) [reply cns/NetworkContainerContract.go:68]: Thanks for catching this, corrected it.

File: cns/NetworkContainerContract.go (hunk 3 above)
Comment (aegal) [reply cns/NetworkContainerContract.go]: Removed

File: cns/NetworkContainerContract.go (hunk 4 above)
Comment (aegal) [reply cns/NetworkContainerContract.go]: Changed.

File: cns/NetworkContainerContract.go (hunk 5 above)
Comment (aegal) [reply cns/NetworkContainerContract.go]: Corrected that.

File: cns/NetworkContainerContract.go (hunk 11 above)
Comment (aegal) [reply cns/NetworkContainerContract.go:240]: Makes sense.

File: cns/NetworkContainerContract.go (hunk 1 above)
Comment (aegal) [reply cns/NetworkContainerContract.go:70]: I think we can leave it as is. Its to me easier to tell between the policies in the request and the static ones we have throughout the code. But will change if you feel otherwise.

File: cns/hnsclient/hnsclient_windows.go (hunk 8 above)
Comment (aegal) [reply cns/hnsclient/hnsclient_windows.go]: Changed

File: cns/hnsclient/hnsclient_windows.go (hunk 9 above)
Comment (aegal) [reply cns/hnsclient/hnsclient_windows.go:516]: Updated

File: cns/hnsclient/hnsclient_windows.go (hunk 10 above)
Comment (aegal) [reply cns/hnsclient/hnsclient_windows.go]: Removed

File: cns/NetworkContainerContract.go (hunk 1 above)
Comment (ashvindeodhar) [reply cns/NetworkContainerContract.go:68]: no, that's fine. thanks

File: cns/NetworkContainerContract.go [hunk 12]
```diff
@@ -220,3 +230,36 @@ type UnpublishNetworkContainerResponse struct {
...
+	RemoteAddresses string `json:","`
+	LocalPorts      string `json:","`
+	RemotePorts     string `json:","`
+	RuleType        string `json:","`
+	Priority        uint16 `json:","`
+}
+
+// Validate - Validates network container request policies
+func (networkContainerRequestPolicy *NetworkContainerRequestPolicies) Validate() error {
+	// validate ACL policy
```
Comment (ashvindeodhar) [inline cns/NetworkContainerContract.go:249]: As one of my comments said can we enforce the validation on apipa endpoint? Check if the provided acl is on apipa endpoint type, if not fail the validation because we don;t have support for acls on cust subnet endpoint. When we add that we should remove the enforced apipa endpoint validation you'll add here. Let me know if it makes sense / you want to discuss this further.

**Comment by ashvindeodhar** [comment]
@tamilmani1989 can you please check the struct to see if this has everything we will need when we extend this to linux as well. If there is anything missing we should add those fields now.
type ValidAclPolicySetting struct {

**Comment by ashvindeodhar** [comment]
@aegal how did you e2e test this?

File: cns/NetworkContainerContract.go (hunk 12 above)
Comment (aegal) [reply cns/NetworkContainerContract.go:249]: Got it, restricted this API only to APIPA in latest commit.
